### PR TITLE
let_users_change_the_code_page

### PR DESCRIPTION
**Pull request by cmcmarrow**
What does this PR do?
Lets cmd/powershell go to utf-8
#57901
#57515

Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by markuskramerIgitt** [comment]
Please remove lines 607-610 that prepend `chcp 437 > nul &` to the command 
```
        else:
            # On Windows set the codepage to US English.
            if python_shell:
                cmd = "chcp 437 > nul & " + cmd

```

image

**Comment by markuskramerIgitt** [comment]
Function chcp() returns a string, chcp_code is an integer.
    
    chcp(chcp_code) != chcp_code

is always true

Please change to 

    chcp(chcp_code) != str(chcp_code)

**Comment by markuskramerIgitt** [comment]
As negative test, I replaced chcp_code=65001 by chcp_code=437 and got the double quotes instead of ä

**Comment by markuskramerIgitt** [comment]
You find my modified cmdmod.py here:
https://github.com/markuskramerIgitt/LearnSalt/tree/master/patch_salt_minion_3001/modules

**Comment by markuskramerIgitt** [comment]
@cmcmarrow looks good for me

Could you please take over my changes, I would like to mass rollout a version we agree upon.

**Comment by cmcmarrow** [comment]
@markuskramerlgitt I’ll take over your changes and add this feature and add it to the startup of salt.

**Review by DmitryKuzmenko — DISMISSED** [review]

**Comment by markuskramerIgitt** [comment]
@cmcmarrow I patched my 3001 Minions with this

File: salt/utils/chcp.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            if raise_error:
+                raise CodePageError()
+            else:
+                log.error("Code page failed to change to %s!", page_id)
+
+    if chcp_ret == "":
+        chcp_ret = chcp()
+
+    log.debug("Code page is %s", chcp_ret)
+    return chcp_ret
```
Comment (twangboy) [inline salt/utils/chcp.py:59]: Can I assume this entire file was replaced by `win_chcp.py`?

Comment (twangboy) [reply salt/utils/chcp.py:59]: And the test with it?

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/utils/chcp.py (hunk 1 above)
Comment (markuskramerIgitt) [reply salt/utils/chcp.py:59]: Yes, @twangboy you assume correctly:    test/chcp.py should be replaced by utils/win_chcp.py
It seems both files exist now, how strange.

Also the test should be replaced, not copied.

Comment (markuskramerIgitt) [reply salt/utils/chcp.py:59]: Yes, @twangboy you assume correctly: test/chcp.py should be replaced by utils/win_chcp.py
It seems both files exist now, how strange.

Also the test should be replaced, not copied.

**Review by twangboy — APPROVED** [review]

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by s0undt3ch** [comment]
This needs to be partially reverted.
It was correct to use `chcp xxx & cmd`, we want to tweak the code page of the command we're going to run, but not necessarily, all other commands after that one.

As for changing code page when salt starts, seen in a comment but not in any code. The admin should be responsible for properly setting his machine, and, if we already support changing the code page to a command we want to run, we're good, by we, I mean salt.

We could check the current code page in order not to chain the `chcp` code page when running the command, but that means an extra shell out per command.

I took a stab at trying to reuse this code, the right way, setting code page, running command, resetting codepage, but that meant 3 shells to run a single command, not good.

Thoughts? Alternatives? Why did we actually needed to shell out to set the code page in the first place and opposed to chaining the commands?

**Comment by s0undt3ch** [comment]
Actually, now that I think about this a bit more, even chaining the chcp command will change the code page for the following commands.....
We should instead do something like `chcp xxx > nul & the actual command & chcp previous_xxx > nul`
How can we grab the current `chcp` in a variable to reset at the end? Using command chaining....

**Comment by twangboy** [comment]
@s0undt3ch There was a lot of discussion about why we did this here: https://github.com/saltstack/salt/pull/57515

**Comment by s0undt3ch** [comment]
Thanks @twangboy, read it.
I still think the merged PR, when setting the codepage, should set it back to how it was.
The system should be properly configured in the first place. At most, our windows bat scripts should do the codepage change at the start of each of our CLIs, not salt.
Anyway, https://github.com/saltstack/salt/pull/58478 seems to correct this behavior, but I'll try testing with the data from the issue you linked.

**Comment by markuskramerIgitt** [comment]
For years Salt was setting code page 437

For year  the codepage, was not set back to how it was.

A change to set it back to how it was should be optional, opt-in, on
request.

I probably sit on the largest minion basis for non 437 Windows boxes, and I
am worried.

I suggest we change the default code page to 437 as it was for years, and
make 65001 an opt-in

**Pull request merged**